### PR TITLE
Updated with Gym wrapper

### PR DESCRIPTION
**Pull request by Shreyas-S-Raman**
Creating gym wrapper for the taxi environment

### CODE REVIEW

**Review by camall3n — CHANGES_REQUESTED** [review]
- What's the `taxi_gym_env_old`? Do we still need it?
- Please delete ".DS_Store" file and configure .gitignore to ignore it in the future

File: taxi/taxi_gym_env.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        taxi_position = state_info[0:2];  info['taxi_x'] = taxi_position[1]; info['taxi_y'] = taxi_position[0]
+
+        #store position + in-taxi information for each passenger
+        for i in range(0,len(state_info[2:]),3):
+
+            p_y, p_x, p_in_taxi = state_info[i:i+3]
+
+            info['p{}_y'.format(i)] = p_y
+            info['p{}_x'.format(i)] = p_x
+            info['p{}_in_taxi'.format(i)] = p_in_taxi
```
Comment (camall3n) [inline taxi/taxi_gym_env.py]: You got the variable order correct, but I'd like to change the labels `'y' -> 'row'`, `'x' -> 'col'` to match the original and be slightly more descriptive.

https://github.com/camall3n/visgrid/blob/d274e6505fb734a84afdde0e04da23b064824927/taxi/taxi.py#L219-L227

Comment (camall3n) [reply taxi/taxi_gym_env.py]: Also, for style purposes, let's have one statement per line, no semicolons.

File: taxi/taxi_gym_env.py [hunk 2]
```diff
@@ -115,15 +115,17 @@ def _get_current_info(self):
         #extract state information for info dict
         state_info = self.taxi_env.get_state()
 
-        taxi_position = state_info[0:2];  info['taxi_x'] = taxi_position[1]; info['taxi_y'] = taxi_position[0]
+        taxi_position = state_info[0:2]
+        info['taxi_col'] = taxi_position[1]
+        info['taxi_row'] = taxi_position[0]
```
Comment (camall3n) [inline taxi/taxi_gym_env.py:124]: Swap these lines

File: taxi/taxi.py [hunk 3]
```diff
@@ -95,7 +96,11 @@ def reset(self, goal=True, explore=False):
         if explore:
             # Fully randomize agent position
             self.agent.position = self.get_random_position()
-
+            
+            #self.agent.position = self.depots[self.passengers[0].color].position
+            #pdb.set_trace()
+            
+
```
Comment (camall3n) [inline taxi/taxi.py:103]: Remove

File: taxi/taxi.py [hunk 4]
```diff
@@ -184,6 +189,8 @@ def render(self):
         return image
 
     def step(self, action):
+        pickup = False; bad_dropoff = False; good_dropoff = False; bad_pickup = False
+
```
Comment (camall3n) [inline taxi/taxi.py:193]:
```
reward_event = None
```

File: taxi/taxi.py [hunk 5]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
                     if (self.agent.position == p.position).all():
                         p.intaxi = True
                         self.passenger = p
+                        pickup = True
```
Comment (camall3n) [inline taxi/taxi.py:207]:
```
reward_event = 'pickup'
```

File: taxi/taxi.py [hunk 6]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
                         break  # max one passenger per taxi
+		
+                if not pickup:
+                    bad_pickup = True
```
Comment (camall3n) [inline taxi/taxi.py:211]:
```
if reward_event != 'pickup':
    reward_event = 'bad_pickup'
```

File: taxi/taxi.py [hunk 7]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
             else:
                 # drop off?
                 dropoff_clear = True
                 for p in (p for p in self.passengers if p is not self.passenger):
                     if (p.position == self.passenger.position).all():
                         dropoff_clear = False
+                        bad_dropoff = True
```
Comment (camall3n) [inline taxi/taxi.py:218]:
```
reward_event = 'bad_dropoff'
```

File: taxi/taxi.py [hunk 8]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
                         break
+
                 if dropoff_clear:
                     self.passenger.intaxi = False
+                    
+		    #drop of current passenger at right depot
```
Comment (camall3n) [inline taxi/taxi.py:224]: Drop *off*

File: taxi/taxi.py [hunk 9]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
+                    depot_position = self.depots[self.passenger.goal].position
+
+                    if np.all(depot_position==self.passenger.position):
+                        good_dropoff = True
```
Comment (camall3n) [inline taxi/taxi.py:228]:
```
reward_event = 'good_dropoff'
```

File: taxi/taxi.py [hunk 10]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
+                    else:
+                        bad_dropoff = True
```
Comment (camall3n) [inline taxi/taxi.py:230]:
```
reward_event = 'bad_dropoff'
```

File: taxi/taxi.py [hunk 11]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
+        #r = 0.0 if not done else 1.0
+        
+        if pickup:
+            r = +5.0
+        elif action==4 and (bad_dropoff or bad_pickup):
+           r = -10.0
+        elif action==4 and good_dropoff:
+            r = +20.0
+        else:
+            r = -1.0
```
Comment (camall3n) [inline taxi/taxi.py:252]: Wrap these different reward functions in an if/else block. The default should be the 0/1 loss, and the taxi constructor can have a new flag `use_gym_rewards=False` that allows for overriding the default reward function.

File: taxi/taxi.py [hunk 12]
```diff
@@ -197,23 +204,59 @@ def step(self, action):
...
+
+        #if done:
+        #    r = +100
+        if r==0:
+            pdb.set_trace()
+            print(s) 
+
```
Comment (camall3n) [inline taxi/taxi.py:259]: Remove

File: taxi/taxi_gym_env.py [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+        self.taxi_env = VisTaxi5x5(n_passengers =  num_passengers, grayscale = grayscale)
+        taxi_rendering = self.taxi_env.reset(goal = have_goal, explore = randomize_positions)
+        taxi_rendering = self._convert_to_int(taxi_rendering)
+
+        #stored for future resets
+        self.have_goal = have_goal; self.randomize_positions = randomize_positions
+
+        #observation space for TaxiEnv: RGB array
+        #self.observation_space = gym.spaces.Box(low=0, high=255, shape = taxi_rendering.shape, dtype = np.uint8)
+        self.observation_space = taxi_rendering
```
Comment (camall3n) [inline taxi/taxi_gym_env.py:30]: What's going on here? Shouldn't the observation_space be a gym `Space` instead of a numpy array?

File: taxi/taxi_gym_env.py (hunk 2 above)
Comment (camall3n) [reply taxi/taxi_gym_env.py:124]: i.e. [0] then [1]

File: utils.py [hunk 14]
```diff
@@ -5,7 +5,7 @@
 import os
 import random
 from sklearn.neighbors import KernelDensity
-import torch
+#import torch
```
Comment (camall3n) [inline utils.py:8]: Remove

**Review by camall3n — CHANGES_REQUESTED** [review]

**Comment by camall3n** [comment]
The latest refactor adds gym support to the base envs (using the newer gym API), so I'm closing this PR.
